File: Core/Auth/FirebaseAuthService.swift
// SPDX-License-Identifier: LicenseRef-Aether3D-Proprietary
// Copyright (c) 2024-2026 Aether3D. All rights reserved.

//
// FirebaseAuthService.swift
// Aether3D
//
// Real AuthService backed by the Firebase iOS SDK. TestFlight / production
// variant: no separate backend; `InternalUserID.rawValue` IS the Firebase
// UID. If we ever add our own backend that mints stable internal ids, the
// change happens inside this file — business code stays on
// `AuthenticatedUser.id` and doesn't notice.
//
// The entire file is wrapped in `#if canImport(FirebaseAuth)` so the target
// still builds cleanly before the user has added the Firebase SPM package.

import Foundation

#if canImport(FirebaseAuth)
import FirebaseAuth
#if canImport(FirebaseCore)
import FirebaseCore
#endif

public final class FirebaseAuthService: AuthService, @unchecked Sendable {

    public init(assumeFirebaseConfigured: Bool = false) {
        if !assumeFirebaseConfigured {
            Self.bootstrapFirebaseIfNeeded()
        }
    }

    // MARK: - AuthService

    public func currentUser() async -> AuthenticatedUser? {
        guard let user = Auth.auth().currentUser else { return nil }
        return Self.toAuthenticatedUser(user)
    }

    public func signIn(_ request: SignInRequest) async throws -> AuthenticatedUser {
        let firebaseUser: User
        switch request {
        case .email(let email, let password):
            firebaseUser = try await mapFirebaseErrors {
                try await Auth.auth().signIn(withEmail: email, password: password).user
            }
        case .phone(_, let verificationID, let code):
            let credential = PhoneAuthProvider.provider().credential(
                withVerificationID: verificationID,
                verificationCode: code
            )
            firebaseUser = try await mapFirebaseErrors {
                try await Auth.auth().signIn(with: credential).user
            }
        }
        return Self.toAuthenticatedUser(firebaseUser)
    }

    public func signUp(_ request: SignUpRequest) async throws -> AuthenticatedUser {
        let firebaseUser: User
        let intendedName: String?
        switch request {
        case .email(let email, let password, let name):
            firebaseUser = try await mapFirebaseErrors {
                try await Auth.auth().createUser(withEmail: email, password: password).user
            }
            intendedName = name
        case .phone(_, let verificationID, let code, let name):
            let credential = PhoneAuthProvider.provider().credential(
                withVerificationID: verificationID,
                verificationCode: code
            )
            firebaseUser = try await mapFirebaseErrors {
                try await Auth.auth().signIn(with: credential).user
            }
            intendedName = name
        }
        if let name = intendedName, !name.isEmpty {
            let change = firebaseUser.createProfileChangeRequest()
            change.displayName = name
            try? await change.commitChanges()
        }
        return Self.toAuthenticatedUser(firebaseUser)
    }

    public func startPhoneVerification(phoneNumber: String) async throws -> PhoneVerificationChallenge {
        let verificationID = try await mapFirebaseErrors {
            try await PhoneAuthProvider.provider().verifyPhoneNumber(phoneNumber, uiDelegate: nil)
        }
        return PhoneVerificationChallenge(
            verificationID: verificationID,
            phoneNumber: phoneNumber,
            expiresAt: Date().addingTimeInterval(5 * 60)
        )
    }

    public func sendEmailVerification() async throws {
        guard let user = Auth.auth().currentUser else {
            throw AuthError.notSignedIn
        }
        try await mapFirebaseErrors {
            try await user.sendEmailVerification()
        }
    }

    public func sendPasswordReset(email: String) async throws {
        try await mapFirebaseErrors {
            try await Auth.auth().sendPasswordReset(withEmail: email)
        }
    }

    public func signOut() async throws {
        try? Auth.auth().signOut()
    }

    public func deleteAccount() async throws {
        if let user = Auth.auth().currentUser {
            try await mapFirebaseErrors {
                try await user.delete()
            }
        }
    }

    // MARK: - Internals

    private static func toAuthenticatedUser(_ user: User) -> AuthenticatedUser {
        return AuthenticatedUser(
            id: InternalUserID(user.uid),
            email: user.email,
            phone: user.phoneNumber,
            displayName: user.displayName
        )
    }

    private func mapFirebaseErrors<T>(_ work: () async throws -> T) async throws -> T {
        do {
            return try await work()
        } catch let error as NSError where error.domain == AuthErrorDomain {
            // Firebase Auth 10.x: `AuthErrorCode` is a struct wrapping the
            // raw Int enum `AuthErrorCode.Code`. Use the .Code form here.
            if let code = AuthErrorCode.Code(rawValue: error.code) {
                throw mapCode(code, underlying: error)
            }
            throw AuthError.unknown(detail: error.localizedDescription)
        } catch {
            throw AuthError.unknown(detail: error.localizedDescription)
        }
    }

    private func mapCode(_ code: AuthErrorCode.Code, underlying: NSError) -> AuthError {
        switch code {
        case .invalidEmail, .wrongPassword, .userNotFound, .invalidCredential:
            return .invalidCredentials
        case .emailAlreadyInUse, .credentialAlreadyInUse:
            return .accountAlreadyExists
        case .weakPassword:
            return .weakPassword
        case .tooManyRequests:
            return .rateLimited
        case .networkError:
            return .network(underlying: underlying.localizedDescription)
        case .invalidVerificationCode, .invalidVerificationID, .sessionExpired:
            return .invalidVerificationCode
        default:
            return .unknown(detail: "firebase_\(code.rawValue):\(underlying.localizedDescription)")
        }
    }

    private static func bootstrapFirebaseIfNeeded() {
        #if canImport(FirebaseCore)
        if FirebaseApp.app() == nil {
            FirebaseApp.configure()
        }
        #endif
    }
}

#endif // canImport(FirebaseAuth)

File: pocketworld_flutter/lib/auth/firebase_auth_service.dart
// Real AuthService backed by Firebase (via firebase_auth pub plugin).

import 'package:firebase_auth/firebase_auth.dart' as fb;

import 'auth_error.dart';
import 'auth_models.dart';
import 'auth_service.dart';

class FirebaseAuthServiceImpl implements AuthService {
  final fb.FirebaseAuth _auth;

  FirebaseAuthServiceImpl({fb.FirebaseAuth? auth})
      : _auth = auth ?? fb.FirebaseAuth.instance;

  @override
  Future<AuthenticatedUser?> currentUser() async {
    final user = _auth.currentUser;
    if (user == null) return null;
    return _toAuthenticatedUser(user);
  }

  @override
  Future<AuthenticatedUser> signIn(SignInRequest request) async {
    return _mapErrors(() async {
      switch (request) {
        case SignInRequestEmail(:final email, :final password):
          final cred = await _auth.signInWithEmailAndPassword(
            email: email,
            password: password,
          );
          final user = cred.user;
          if (user == null) throw const AuthException(AuthErrorKind.unknown);
          return _toAuthenticatedUser(user);
        case SignInRequestPhone(:final verificationID, :final code):
          final phoneCred = fb.PhoneAuthProvider.credential(
            verificationId: verificationID,
            smsCode: code,
          );
          final cred = await _auth.signInWithCredential(phoneCred);
          final user = cred.user;
          if (user == null) throw const AuthException(AuthErrorKind.unknown);
          return _toAuthenticatedUser(user);
      }
    });
  }

  @override
  Future<AuthenticatedUser> signUp(SignUpRequest request) async {
    return _mapErrors(() async {
      late fb.User firebaseUser;
      String? intendedName;
      switch (request) {
        case SignUpRequestEmail(
            :final email,
            :final password,
            :final displayName
          ):
          final cred = await _auth.createUserWithEmailAndPassword(
            email: email,
            password: password,
          );
          final user = cred.user;
          if (user == null) throw const AuthException(AuthErrorKind.unknown);
          firebaseUser = user;
          intendedName = displayName;
        case SignUpRequestPhone(
            :final verificationID,
            :final code,
            :final displayName
          ):
          final phoneCred = fb.PhoneAuthProvider.credential(
            verificationId: verificationID,
            smsCode: code,
          );
          final cred = await _auth.signInWithCredential(phoneCred);
          final user = cred.user;
          if (user == null) throw const AuthException(AuthErrorKind.unknown);
          firebaseUser = user;
          intendedName = displayName;
      }
      if (intendedName != null && intendedName.isNotEmpty) {
        try {
          await firebaseUser.updateDisplayName(intendedName);
        } catch (_) {}
      }
      return _toAuthenticatedUser(firebaseUser);
    });
  }

  @override
  Future<PhoneVerificationChallenge> startPhoneVerification(
    String phoneNumber,
  ) async {
    final completer = _PhoneVerificationCompleter();
    await _mapErrors(() async {
      await _auth.verifyPhoneNumber(
        phoneNumber: phoneNumber,
        timeout: const Duration(minutes: 2),
        verificationCompleted: (_) {},
        verificationFailed: (e) {
          completer.completeError(_mapFirebaseAuthException(e));
        },
        codeSent: (verificationId, _) {
          completer.complete(PhoneVerificationChallenge(
            verificationID: verificationId,
            phoneNumber: phoneNumber,
            expiresAt: DateTime.now().add(const Duration(minutes: 5)),
          ));
        },
        codeAutoRetrievalTimeout: (verificationId) {
          if (!completer.isCompleted) {
            completer.complete(PhoneVerificationChallenge(
              verificationID: verificationId,
              phoneNumber: phoneNumber,
              expiresAt: DateTime.now().add(const Duration(minutes: 5)),
            ));
          }
        },
      );
    });
    return completer.future;
  }

  @override
  Future<void> sendEmailVerification() async {
    final user = _auth.currentUser;
    if (user == null) {
      throw const AuthException(AuthErrorKind.notSignedIn);
    }
    await _mapErrors(() => user.sendEmailVerification());
  }

  @override
  Future<void> sendPasswordReset(String email) async {
    await _mapErrors(() => _auth.sendPasswordResetEmail(email: email));
  }

  @override
  Future<void> signOut() async {
    try {
      await _auth.signOut();
    } catch (_) {}
  }

  @override
  Future<void> deleteAccount() async {
    final user = _auth.currentUser;
    if (user == null) return;
    await _mapErrors(() => user.delete());
  }

  AuthenticatedUser _toAuthenticatedUser(fb.User user) {
    return AuthenticatedUser(
      id: InternalUserID(user.uid),
      email: user.email,
      phone: user.phoneNumber,
      displayName: user.displayName,
    );
  }

  Future<T> _mapErrors<T>(Future<T> Function() work) async {
    try {
      return await work();
    } on fb.FirebaseAuthException catch (e) {
      throw _mapFirebaseAuthException(e);
    } on AuthException {
      rethrow;
    } catch (e) {
      throw AuthException(AuthErrorKind.unknown, e.toString());
    }
  }

  AuthException _mapFirebaseAuthException(fb.FirebaseAuthException e) {
    switch (e.code) {
      case 'invalid-email':
      case 'wrong-password':
      case 'user-not-found':
      case 'invalid-credential':
        return AuthException(AuthErrorKind.invalidCredentials, e.code);
      case 'email-already-in-use':
      case 'credential-already-in-use':
        return AuthException(AuthErrorKind.accountAlreadyExists, e.code);
      case 'weak-password':
        return AuthException(AuthErrorKind.weakPassword, e.code);
      case 'too-many-requests':
        return AuthException(AuthErrorKind.rateLimited, e.code);
      case 'network-request-failed':
        return AuthException(AuthErrorKind.network, e.code);
      case 'invalid-verification-code':
      case 'invalid-verification-id':
      case 'session-expired':
        return AuthException(AuthErrorKind.invalidVerificationCode, e.code);
      default:
        return AuthException(
          AuthErrorKind.unknown,
          'firebase_${e.code}:${e.message ?? ''}',
        );
    }
  }
}

class _PhoneVerificationCompleter {
  bool _completed = false;
  final _inner = <PhoneVerificationChallenge>[];
  final _errors = <Object>[];

  void complete(PhoneVerificationChallenge v) {
    if (_completed) return;
    _completed = true;
    _inner.add(v);
  }

  void completeError(Object e) {
    if (_completed) return;
    _completed = true;
    _errors.add(e);
  }

  bool get isCompleted => _completed;

  Future<PhoneVerificationChallenge> get future async {
    while (!_completed) {
      await Future<void>.delayed(const Duration(milliseconds: 10));
    }
    if (_errors.isNotEmpty) throw _errors.first;
    return _inner.first;
  }
}
